packages/dropdown/src/react/index.js

```diff
@@ -157,9 +158,33 @@ const Dropdown = React.forwardRef((props, forwardedRef) => {
       setKeyboarding(false)
     }
   }
+
   useEffect(() => {
-    isOpen && document.addEventListener('click', handleClickOutsideMenu)
-    return () => document.removeEventListener('click', handleClickOutsideMenu)
+    let currentAnimationFrame = null
+    const requestAnimationFrame = () => {
+      window.cancelAnimationFrame(currentAnimationFrame)
+      currentAnimationFrame = window.requestAnimationFrame(() => setOpen(false))
+      return currentAnimationFrame
+    }
+    let timeout = null
+    const clear = () => {
+      clearTimeout(timeout)
+      timeout = setTimeout(() => setOpen(false), 50)
+    }
+    if (isOpen) {
+      document.addEventListener('click', handleClickOutsideMenu)
+      document.addEventListener('contextmenu', handleClickOutsideMenu)
+      window.addEventListener('resize', requestAnimationFrame, {
+        passive: true
+      })
+      window.addEventListener('scroll', clear, { passive: true })
+    }
+    return () => {
+      document.removeEventListener('click', handleClickOutsideMenu)
+      document.removeEventListener('contextmenu', handleClickOutsideMenu)
+      window.removeEventListener('resize', requestAnimationFrame)
+      window.removeEventListener('scroll', clear)
```

Comment:
Update conditions on which we close the dropdown automatically. In addition to **click out side**, we now have:
 - **right click outside**
- **scroll event**
- **resize event**

Reply:
i know you mentioned in standup you felt this hook was too large. one simple change could be separating it into multiple hooks. that would also make the code more readable and the hooks would be reusable. here is some code from the flow extensions repo that illustrates how you could organize the hooks into smaller and more reusable chunks

```
import { canUseEventListeners, canUseDOM } from 'exenv';
import { useEffect, useRef } from 'react';

const ESCAPE_KEY = 27;

export function useGlobalEventListener<EventName extends keyof WindowEventMap>(name: EventName, fn: (evt: WindowEventMap[EventName]) => unknown): void {
    const handlerRef = useRef<typeof fn>();

    useEffect(() => {
        handlerRef.current = fn;
    }, [fn]);

    useEffect(() => {
        if (!canUseEventListeners || !canUseDOM) return;

        const handler = handlerRef.current!;

        window.addEventListener(name, handler, true);

        return () => {
            window.removeEventListener(name, handler, true);
        };
    }, [name]);
}

export function useOnDocumentClick(handler: (evt: any) => unknown) {
    useGlobalEventListener('mousedown', handler);
    useGlobalEventListener('touchstart', handler);
}

export function useOnEscape(handler: (evt: KeyboardEvent) => unknown) {
    function handleKeyDown(evt: KeyboardEvent) {
        if (evt.keyCode !== ESCAPE_KEY) return;
        handler(evt);
    }

    useGlobalEventListener('keydown', handler);
}
```

Reply:
I'll update the PR to use this pattern

Reply:
Going to pass on this suggestion. Their is a conditional requirement for these functions making the pattern above more complex in it's implementation than I think is necessary.